#**Bank Marketing Dataset**
####By: Jessica, Carmen, Yanting and Emma
#### Anonymous bank located in Portugal
#### The business objective is to identify which customer characteristics are associated with a higher likelihood of subscribing to a term deposit so that marketing efforts can be focused on the most promising prospects.

In [ ]:
# Import Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# Load CSV
from google.colab import files
upload = files.upload()

Saving bank-direct-marketing-campaigns.csv to bank-direct-marketing-campaigns (10).csv


In [ ]:
# Read CSV
df = pd.read_csv("bank-direct-marketing-campaigns.csv")
df

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
41183,73,retired,married,professional.course,no,yes,no,cellular,nov,fri,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes
41184,46,blue-collar,married,professional.course,no,no,no,cellular,nov,fri,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41185,56,retired,married,university.degree,no,yes,no,cellular,nov,fri,2,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41186,44,technician,married,professional.course,no,no,no,cellular,nov,fri,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes


### Feature descriptions

##### Age, the category of job (admin, blue-collar, student, etc), marital status, education level, default (if they have unpaid debt), housing loan, loan (personal), contact (communication type), month (last contact month), day of week (last contact day of week), campaign (number of contacts performed), pdays (number of days after client last contacted), previous (number of contacts before this campaign), poutcome (outcome of previous campaign), emp.var.rate (emplayment variation rate), cons.price.idx (consumer price index), cons.conf.idx (consumer confidence index), euribor3m (euribor 3 month rate), nr.employed (number of employees), y (has client subscribed a term deposit?)

In [ ]:
#shape
df.shape

(41188, 20)

In [ ]:
# data info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  campaign        41188 non-null  int64  
 11  pdays           41188 non-null  int64  
 12  previous        41188 non-null  int64  
 13  poutcome        41188 non-null  object 
 14  emp.var.rate    41188 non-null  float64
 15  cons.price.idx  41188 non-null  float64
 16  cons.conf.idx   41188 non-null  float64
 17  euribor3m       41188 non-null 

In [ ]:
df.isna().sum()

,0
age,0
job,0
marital,0
education,0
default,0
housing,0
loan,0
contact,0
month,0
day_of_week,0


In [ ]:
# Drop rows where y is missing
df = df.dropna(subset=['y'])

In [ ]:
# Count the number of data that says unknown
unknown_counts = (df == "unknown").sum()

print(unknown_counts[unknown_counts > 0])

job           330
marital        80
education    1731
default      8597
housing       990
loan          990
dtype: int64


In [ ]:
# Percentage of data that is unknown
unknown_percent = ((df == "unknown").mean() * 100).round(2)

print(unknown_percent[unknown_percent > 0])

job           0.80
marital       0.19
education     4.20
default      20.87
housing       2.40
loan          2.40
dtype: float64


In [ ]:
# Describe the dataset
df.describe()

,age,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed
count,41188.00000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000
mean,40.02406,2.567593,962.475454,0.172963,0.081886,93.575664,-40.502600,3.621291,5167.035911
std,10.42125,2.770014,186.910907,0.494901,1.570960,0.578840,4.628198,1.734447,72.251528
min,17.00000,1.000000,0.000000,0.000000,-3.400000,92.201000,-50.800000,0.634000,4963.600000
25%,32.00000,1.000000,999.000000,0.000000,-1.800000,93.075000,-42.700000,1.344000,5099.100000
50%,38.00000,2.000000,999.000000,0.000000,1.100000,93.749000,-41.800000,4.857000,5191.000000
75%,47.00000,3.000000,999.000000,0.000000,1.400000,93.994000,-36.400000,4.961000,5228.100000
max,98.00000,56.000000,999.000000,7.000000,1.400000,94.767000,-26.900000,5.045000,5228.100000


These are all of the features that were numerical prior to changing them to
binary or scaling them. By describing the data it gave us a better
understanding of how it lies before we get into cleaning it.

In [ ]:
# Professional descriptive summary
descriptive_summary = df.describe().T
descriptive_summary

,count,mean,std,min,25%,50%,75%,max
age,41188.0,40.024060,10.421250,17.000,32.000,38.000,47.000,98.000
campaign,41188.0,2.567593,2.770014,1.000,1.000,2.000,3.000,56.000
pdays,41188.0,962.475454,186.910907,0.000,999.000,999.000,999.000,999.000
previous,41188.0,0.172963,0.494901,0.000,0.000,0.000,0.000,7.000
emp.var.rate,41188.0,0.081886,1.570960,-3.400,-1.800,1.100,1.400,1.400
cons.price.idx,41188.0,93.575664,0.578840,92.201,93.075,93.749,93.994,94.767
cons.conf.idx,41188.0,-40.502600,4.628198,-50.800,-42.700,-41.800,-36.400,-26.900
euribor3m,41188.0,3.621291,1.734447,0.634,1.344,4.857,4.961,5.045
nr.employed,41188.0,5167.035911,72.251528,4963.600,5099.100,5191.000,5228.100,5228.100


In [ ]:
# Check for duplicate rows
print("Number of duplicated rows:", df.duplicated().sum())

Number of duplicated rows: 1784


In [ ]:
# Delete duplicate rows
df = df.drop_duplicates()
print("New dataset shape:")
print(df.shape)

New dataset shape:
(39404, 20)


Checked for duplicate rows and found that there were 1784 of them. So,
we decided to drop them.

In [ ]:
#Checking
cols_to_check = ['age', 'campaign']

# Keep a record of your original row count
original_rows = len(df)

for col in cols_to_check:
    # Calculate Q1 (25th percentile) and Q3 (75th percentile)
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)

    # Calculate the IQR range
    IQR = Q3 - Q1

    # Define your upper and lower boundaries
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Filter the dataframe to only keep rows within the boundaries
    df = df[(df[col] >= lower_bound) & (df[col] <= upper_bound)]

print(f"Dropped {original_rows - len(df)} statistical outlier rows.")

Dropped 2842 statistical outlier rows.


In [ ]:
# check for outliers
df = check_outliers(df, ['age'], 5)

In [ ]:
df[df.outlier_age]

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,outlier_age


In [ ]:
df = df.drop('outlier_age', axis=1)

There were outliars in both the age and campaign section of out data so we dropped those to prevent it from becoming skewed

In [ ]:
df.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


In [ ]:
# Change catrgorical to binary
df = pd.get_dummies(
    data=df,
    columns=['job', 'education','marital','default','housing','loan','contact','month','day_of_week','y','poutcome'],
    drop_first=True
)

In [ ]:
df.head()

,age,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,job_blue-collar,...,month_nov,month_oct,month_sep,day_of_week_mon,day_of_week_thu,day_of_week_tue,day_of_week_wed,y_yes,poutcome_nonexistent,poutcome_success
0,56,1,999,0,1.1,93.994,-36.4,4.857,5191.0,False,...,False,False,False,True,False,False,False,False,True,False
1,57,1,999,0,1.1,93.994,-36.4,4.857,5191.0,False,...,False,False,False,True,False,False,False,False,True,False
2,37,1,999,0,1.1,93.994,-36.4,4.857,5191.0,False,...,False,False,False,True,False,False,False,False,True,False
3,40,1,999,0,1.1,93.994,-36.4,4.857,5191.0,False,...,False,False,False,True,False,False,False,False,True,False
4,56,1,999,0,1.1,93.994,-36.4,4.857,5191.0,False,...,False,False,False,True,False,False,False,False,True,False


In [ ]:
bool_cols = df.select_dtypes(include='bool').columns

df[bool_cols] = df[bool_cols].astype(int)

In [ ]:
df.head()

,age,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,job_blue-collar,...,month_nov,month_oct,month_sep,day_of_week_mon,day_of_week_thu,day_of_week_tue,day_of_week_wed,y_yes,poutcome_nonexistent,poutcome_success
0,56,1,999,0,1.1,93.994,-36.4,4.857,5191.0,0,...,0,0,0,1,0,0,0,0,1,0
1,57,1,999,0,1.1,93.994,-36.4,4.857,5191.0,0,...,0,0,0,1,0,0,0,0,1,0
2,37,1,999,0,1.1,93.994,-36.4,4.857,5191.0,0,...,0,0,0,1,0,0,0,0,1,0
3,40,1,999,0,1.1,93.994,-36.4,4.857,5191.0,0,...,0,0,0,1,0,0,0,0,1,0
4,56,1,999,0,1.1,93.994,-36.4,4.857,5191.0,0,...,0,0,0,1,0,0,0,0,1,0


By converting our categorical values into binary we are able to easily inpterpat and compare finding in duture analysis

In [ ]:
# Scale appropriate columns
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler(feature_range=(-1, 1))

cols_to_scale = [

    'campaign','previous','emp.var.rate','cons.price.idx','nr.employed','pdays','cons.conf.idx', 'euribor3m','age']

existing_cols = [c for c in cols_to_scale if c in df.columns]

df[existing_cols] = scaler.fit_transform(df[existing_cols])

df.head()

,age,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,job_blue-collar,...,month_nov,month_oct,month_sep,day_of_week_mon,day_of_week_thu,day_of_week_tue,day_of_week_wed,y_yes,poutcome_nonexistent,poutcome_success
0,0.500000,-1.0,1.0,-1.0,0.875,0.397506,0.205021,0.914759,0.719471,0,...,0,0,0,1,0,0,0,0,1,0
1,0.538462,-1.0,1.0,-1.0,0.875,0.397506,0.205021,0.914759,0.719471,0,...,0,0,0,1,0,0,0,0,1,0
2,-0.230769,-1.0,1.0,-1.0,0.875,0.397506,0.205021,0.914759,0.719471,0,...,0,0,0,1,0,0,0,0,1,0
3,-0.115385,-1.0,1.0,-1.0,0.875,0.397506,0.205021,0.914759,0.719471,0,...,0,0,0,1,0,0,0,0,1,0
4,0.500000,-1.0,1.0,-1.0,0.875,0.397506,0.205021,0.914759,0.719471,0,...,0,0,0,1,0,0,0,0,1,0


We scaled the remaining numerical values for easy compairsion and so that no values would be weighted more than others withing the compairions.

In [ ]:
# download as CSV.
output_file_path = 'bank_marketing_cleaned_for_classification.csv'
df.to_csv(output_file_path, index=False)